File: src/main/scala/com/abaddon83/cardsGames/burracoGames/domainModels/burracoGames/initialised/mazzettos/MazzettoDecks.scala
package com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.mazzettos

import com.abaddon83.cardsGames.shares.decks.Card

case class MazzettoDecks protected(protected val list: List[MazzettoDeck]){

  def firstMazzettoAvailable(): MazzettoDeck = {
    assert(!list.isEmpty, "Mazzetto list empty, all Mazzetto taken")
    list.head
  }

  def mazzettoTaken(mazzettoDeck: MazzettoDeck): MazzettoDecks = {
    assert(list.exists(m => m == mazzettoDeck), "MazzettoDeck not found")
    copy(
      list = list diff List(mazzettoDeck)
    )
  }

  def numCards(): Int ={
    list.map(m => m.numCards()).foldLeft(0)(_ + _)
  }
}

object MazzettoDecks{
  def build(list: List[MazzettoDeck]): MazzettoDecks = {
    assert(list.size == 2,"MazzettoDecks can accept only 2 MazzettoDecks")
    MazzettoDecks(list.sortBy(p=> p.numCards()).reverse)
  }
}

File: src/main/kotlin/com/abaddon83/burracoGame/domainModels/MazzettoDecks.kt
package com.abaddon83.burracoGame.domainModels


data class MazzettoDecks private constructor(val list: List<MazzettoDeck>) {

    fun firstMazzettoAvailable(): MazzettoDeck {
        assert(list.isNotEmpty()){"Mazzetto list empty, all Mazzetto taken"}
        return list.first()
    }

    fun mazzettoTaken(mazzettoDeck: MazzettoDeck): MazzettoDecks {
        assert(list.find{m -> m == mazzettoDeck}!= null) {"MazzettoDeck not found"}
        return copy(list = list.minus(mazzettoDeck))
    }

    fun numCards(): Int = list.map{m -> m.numCards()}.fold(0){ total, item -> total + item }

    companion object Factory {
        fun create(list: List<MazzettoDeck>): MazzettoDecks {
            assert(list.size == 2) {"MazzettoDecks can accept only 2 MazzettoDecks"}
            return MazzettoDecks(list.sortedBy {p-> p.numCards()}.reversed())
        }
    }
}